hw/ip/otbn/util/check_loop.py

```diff
@@ -0,0 +1,250 @@
+
+    The code section is considered to include both the start and end PC.
+    '''
+    def __init__(self, start: int, end: int):
+        self.start = start
+        self.end = end
+
+    def __contains__(self, pc):
+        return self.start <= pc and pc <= self.end
+
+    def __repr__(self):
+        return '0x{:x}-0x{:x}'.format(self.start, self.end)
+
+
+def _get_pcs_for_mnemonic(program: OTBNProgram, mnem: str) -> List[int]:
+    '''Gets all PCs in the program at which the given instruction is present.'''
+    imem_start = program.get_pc_at_symbol('_imem_start')
+    imem_end = program.get_pc_at_symbol('_imem_end')
+    out = []
+    for pc in range(imem_start, imem_end, 4):
+        insn = program.get_insn(pc)
+        if insn.mnemonic == mnem:
+            out.append(pc)
+    return out
+
+
+def _get_branches(program: OTBNProgram) -> List[int]:
+    '''Gets the PCs of all branch instructions (BEQ and BNE) in the program.'''
+    return _get_pcs_for_mnemonic(program, 'bne') + _get_pcs_for_mnemonic(
+        program, 'beq')
+
+
+def _get_loop_starts(program: OTBNProgram) -> List[int]:
+    '''Gets the start PCs of all loops (LOOP and LOOPI) in the program.'''
+    return _get_pcs_for_mnemonic(program, 'loop') + _get_pcs_for_mnemonic(
+        program, 'loopi')
+
+
+def _get_loops(program: OTBNProgram) -> List[CodeSection]:
+    '''Gets the PC ranges of all loops (LOOP and LOOPI) in the program.'''
+    loop_starts = _get_loop_starts(program)
+    loops = []
+    for pc in loop_starts:
+        insn = program.get_insn(pc)
+        operands = program.get_operands(pc)
+        end_pc = pc + operands['bodysize'] * 4
+        loops.append(CodeSection(pc + 4, end_pc))
+    return loops
+
+
+def _check_loop_iterations(program: OTBNProgram,
+                           loops: List[CodeSection]) -> CheckResult:
+    '''Checks number of iterations for loopi.
+
+    If the number of iterations is 0, this check fails; `loopi` requires at least
+    one iteration and will raise a LOOP error otherwise. The `loop` instruction
+    also has this requirement, but since the number of loop iterations comes
+    from a register it's harder to check statically and is not considered here.
+    '''
+    out = CheckResult()
+    for l in loops:
+        insn = program.get_insn(l.start)
+        operands = program.get_operands(l.start)
+        if insn.mnemonic == 'loopi' and operands['iterations'] <= 0:
+            out.err(
+                'Bad number of loop iterations ({}) at PC 0x{:#x}: {}'.format(
+                    operands['iterations'], l.start,
+                    insn.disassemble(l.start, operands)))
+    return out
+
+
+def _insn_ok_at_loop_end(insn: Insn) -> bool:
+    if insn.errs is None:
+        return True
+    for err in insn.errs:
+        if 'LOOP' not in err:
+            continue
+        if err == 'A LOOP error if this instruction appears as the last instruction of a loop body.':
```

Comment:
Maybe we should use constants for this instead of stringly-typed errors?

Reply:
I wouldn't even do that. I'd suggest inlining the knowledge of which instructions can't appear at the end of a loop (jumps, branches, loops). It's spread around all over the tools and isn't going to change.

Reply:
I'll change this to just check the `straight-line` property, which is false for control flow instructions; seems like a better way to do it in retrospect.